File: rust/src/phone.rs
const MAX:usize = 10;
struct Node {
    flags: u64,
    iso: String,
    child: [Option<Box<Node>>; MAX]
}

static mut TREE: [Option<Box<Node>>; MAX] = [None,None,None,None,None,None,None,None,None,None];

pub fn add(nr: &str, iso: &str) {
    let mut tree = unsafe { &mut TREE };
    for (i, ch) in nr.chars().enumerate() {
        let index = ch.to_digit(10).unwrap() as usize;
        let value = if i == nr.len() - 1 {1} else {0};
        if tree[index].as_mut().is_none() {
            let node = tree[index].get_or_insert( Box::<Node>::new(
                Node {
                    flags: value,
                    iso: if value != 0 {iso.to_string()} else {"".to_owned()},
                    child: Default::default()
                }));
            tree = &mut node.child;
        } else {
            let node = tree[index].as_mut().unwrap().as_mut();
            if value != 0 {
                node.flags = value;
                node.iso = iso.to_string();
            }
            tree = &mut node.child;
        }
    }
 }

pub fn parse(nr: &str) -> (u8, u8, u8, &str) {
    let mut len = 0u8;
    let mut ld = 0u8;
    let mut ac = 0u8;
    let mut iso: &str = "";

    let mut tree = unsafe { &TREE };
    for (i, ch) in nr.chars().enumerate() {
        let index = ch.to_digit(10).unwrap() as usize;
        let opt = tree[index].as_ref();
        if opt.is_none() { break; }
        let node = opt.unwrap().as_ref();
        if node.flags == 1 {
            if len == 0 { len = i as u8 + 1; }
            else if ld == 0 { ld = i as u8 + 1; }
            else { ac = i as u8 + 1; }
            iso = &node.iso;
        }
        tree = &node.child;
    }

    return (len,ld,ac,iso);
} 

File: dart/lib/phone.dart

class _Node {
  int value = 0;
  String iso;
  List<_Node> child = List(10);
}

var _TREE = _Node();

int add(String cc, String iso) {
  var tree = _TREE;
  for (var i = 0; i < cc.length; ++i) {
    var d = cc.codeUnitAt(i) - '0'.codeUnitAt(0);
    if (tree.child[d] == null) {
      tree.child[d] = _Node();
    }
    var node = tree.child[d];
    if (i == cc.length - 1) {
      node.value = 1;
      node.iso = iso;
    }
    tree = node;
  }
  return 0;
}

class PhoneInfo {
  int cc = 0;
  int ld = 0;
  int ac = 0;
  String iso;
}

PhoneInfo match(String phone) {
  var tree = _TREE;
  var info = PhoneInfo();
  for (var i = 0; i < phone.length && tree != null; ++i) {
    var d = phone.codeUnitAt(i) - '0'.codeUnitAt(0);
    var node = tree.child[d];
    if (node != null && node.value == 1) {
      if (info.cc == 0) {
        info.cc = i + 1;
      } else if (info.ld == 0) {
        info.ld = i + 1;
      } else {
        info.ac = i + 1;
      }
      info.iso = node.iso;
    } 
    tree = node;
  }
  return info;
}

int calculate() {
  return 6 * 7;
}